# 03장. TF-IDF — 글자를 숫자로

| 오늘의 질문 | 예상 시간 |
|---|---:|
| 컴퓨터는 메뉴 글자를 어떻게 숫자로 바꿀까? | 3회차 전반 · 약 90분 |


## 이 장에서 배울 내용

- 한 단어를 2-gram과 3-gram으로 직접 나눌 수 있다.
- TF, DF, IDF가 각각 무엇을 세는지 표로 설명할 수 있다.
- TF와 IDF를 곱해 한 글자 조각의 TF-IDF를 계산할 수 있다.
- 출처가 기록된 한국어 문서 세 편에서 문서별 특징어를 찾을 수 있다.
- 취향 문장과 메뉴를 숫자로 비교할 준비를 할 수 있다.


## 생각 열기

‘토마토파스타를 좋아한다’라는 취향과 ‘미트볼파스타’라는 메뉴는 비슷해 보입니다. 그러나 컴퓨터는 문장의 느낌을 그대로 비교하지 못합니다. 두 문장에 함께 나타나는 글자 조각을 세어 숫자로 표현해 봅시다.


## 핵심 용어

| 용어 | 뜻 |
|---|---|
| **n-gram** | 연속된 n개의 글자 조각 |
| **TF** | 한 문서 안에서 글자 조각이 나타난 정도 |
| **DF** | 그 글자 조각이 들어 있는 문서의 수 |
| **IDF** | 여러 문서 중 드물게 나타나 구별에 도움 되는 정도 |
| **TF-IDF** | 한 문서 안의 중요도와 전체 문서에서의 희소성을 곱한 값 |
| **말뭉치** | 같은 방법으로 비교하려고 모아 둔 여러 문서 |
| **벡터** | 여러 숫자를 순서대로 모은 표현 |


## 개념 익히기


### 1. n-gram은 움직이는 창으로 만드는 글자 조각

컴퓨터는 문장을 뜻으로 바로 이해하지 못하므로 먼저 비교할 수 있는 작은 조각을 만듭니다. n-gram의 n은 **한 조각에 들어가는 글자 수**입니다.

`파스타` 위에 두 칸짜리 창을 올리고 한 칸씩 옮기면 `파스`, `스타`가 나옵니다. 이것이 2-gram입니다. 세 칸짜리 창을 쓰면 `파스타` 하나가 나오며, 이것은 3-gram입니다.

이 프로젝트는 단어 앞뒤의 경계도 구별하려고 공백을 붙입니다. 눈에 보이지 않는 공백을 `·`로 표시하면 `파스타`의 2-gram은 `·파`, `파스`, `스타`, `타·`입니다. 2글자 조각은 짧은 공통점을 잘 찾고, 3~4글자 조각은 더 구체적인 이름을 잡습니다. 그래서 한 크기만 고르지 않고 2~4글자를 함께 사용합니다.

한국어 메뉴는 `치즈 파스타`, `치즈파스타`처럼 띄어쓰기가 달라질 수 있습니다. 단어 전체가 정확히 같아야 하는 방식보다 문자 조각 방식이 작은 표기 차이에서도 공통 부분을 찾기 쉽습니다.

### 2. TF는 한 문장 안을 본다

TF는 **Term Frequency**, 즉 한 문장 안에서 특정 조각이 차지하는 비율입니다.

    TF = 그 조각이 나온 횟수 ÷ 그 문장의 전체 n-gram 수

같은 조각이 많이 나오면 TF가 커집니다. 단순 횟수 대신 전체 조각 수로 나누는 까닭은 짧은 메뉴와 긴 메뉴를 조금 더 공정하게 비교하기 위해서입니다.

### 3. DF와 IDF는 모든 문장을 함께 본다

DF는 **Document Frequency**입니다. 어떤 조각이 총 몇 번 반복됐는지가 아니라, 그 조각을 포함한 문서가 몇 개인지를 셉니다. 한 문서에서 열 번 반복되어도 DF에는 문서 한 개로 셉니다.

반 학생 모두가 ‘급식’이라는 말을 쓰면 그 말만으로 누가 쓴 문장인지 구별하기 어렵습니다. 반대로 한 학생만 ‘파스타’라는 말을 썼다면 강한 단서가 됩니다. IDF는 이런 차이를 숫자로 나타냅니다.

    IDF = log((전체 문서 수 + 1) ÷ (DF + 1)) + 1

DF가 작으면 나누는 수가 작아져 IDF가 커지고, DF가 크면 IDF가 작아집니다. `log`는 드문 조각의 값이 지나치게 커지지 않도록 눌러 줍니다. 식의 `+1`은 문서가 적거나 처음 보는 조각이 있어도 안전하게 계산하기 위한 장치입니다.

### 4. TF-IDF는 두 관점을 합친다

    TF-IDF = TF × IDF

어떤 조각이 **이 메뉴 안에서는 자주 나오고(TF가 큼)**, **다른 메뉴에는 드물면(IDF가 큼)** TF-IDF가 커집니다. 모든 조각의 TF-IDF 값을 같은 순서로 늘어놓은 숫자 목록이 벡터입니다. 다음 장에서는 두 벡터가 가리키는 방향을 비교합니다.

### 5. 실제 문서는 먼저 같은 기준으로 단어를 골라야 한다

문서 한 편만 읽으면 어떤 단어가 그 글만의 특징인지 판단하기 어렵습니다. 주제가 다른 여러 문서를 한데 모은 **말뭉치**에서 TF와 IDF를 함께 계산해야 합니다. 이 장에서는 공개 라이선스로 배포된 한국어 AI 입문 문서 세 편을 직접 파일로 읽습니다.

한국어는 `이미지`, `이미지를`, `이미지는`처럼 조사가 붙습니다. 전문 형태소 분석기를 쓰면 이를 더 정교하게 나눌 수 있지만, 이번 활동에서는 원리를 눈으로 확인하는 것이 먼저입니다. `[가-힣]{2,}`라는 규칙으로 한글 두 글자 이상을 찾고, `있습니다`, `그리고`처럼 주제를 구별하기 어려운 자주 쓰는 말은 작은 불용어 목록에서 뺍니다. 따라서 결과는 완벽한 국어 분석이 아니라 **같은 단순 규칙으로 세 문서를 공정하게 비교한 결과**입니다.

### 이 장의 TF-IDF 튜토리얼 순서

공식만 읽고 넘어가지 않습니다. 아래 여섯 단계를 차례로 실행하면서 **어느 문서에서 어떤 단어가 몇 점을 받았는지** 끝까지 추적합니다.

1. **1단계. 비교할 문서 세 편 확인하기** — 제목, 글자 수, 출처를 확인합니다.
2. **2단계. 문서에서 분석할 한글 단어 고르기** — 단어를 뽑고 흔한 표현을 제외합니다.
3. **3단계. 단어 하나의 계산 과정 따라가기** — 등장 횟수, 전체 단어 수, TF, DF, IDF, TF-IDF를 차례로 계산합니다.
4. **4단계. 문서별 상위 TF-IDF 결과 비교하기** — 세 문서의 상위 단어와 실제 점수를 한 표에서 봅니다.
5. **5단계. 결과를 문서 내용과 연결해 해석하기** — 점수가 높은 까닭과 분석의 한계를 문장으로 설명합니다.
6. **6단계. 같은 방법을 급식 메뉴에 적용하기** — 가상 취향과 다섯 메뉴의 순위, 유사도, 공통 글자 조각 수를 확인합니다.

표의 소수는 읽기 쉽도록 반올림합니다. 계산에는 반올림하기 전 값을 사용하므로 손으로 계산한 값과 마지막 자리에서 조금 다를 수 있습니다.


## 활동 전 생각


종이에 `파스타`를 쓰고 단어 앞뒤에 경계 표시 `·`를 붙여 `·파스타·`로 만드세요.

| 창 크기 | 창을 한 칸씩 옮긴 결과 |
|---|---|
| 2-gram | `·파`, `파스`, `스타`, `타·` |
| 3-gram | `·파스`, `파스타`, `스타·` |

이제 A=`파스타 피자`, B=`파스타`, C=`밥 국` 세 문서에서 `파스`와 `피자`를 비교합니다. 추천기에 쓰는 2~4글자 조각을 모두 세면 A는 15개, B는 9개, C는 6개입니다.

| 문서 | 조각 | 등장 횟수 | 전체 조각 수 | TF | DF | IDF | TF-IDF |
|---|---|---:|---:|---:|---:|---:|---:|
| A | `파스` | 1 | 15 | 0.067 | 2 | 1.288 | 0.086 |
| A | `피자` | 1 | 15 | 0.067 | 1 | 1.693 | 0.113 |
| B | `파스` | 1 | 9 | 0.111 | 2 | 1.288 | 0.143 |
| B | `피자` | 0 | 9 | 0.000 | 1 | 1.693 | 0.000 |
| C | `파스` | 0 | 6 | 0.000 | 2 | 1.288 | 0.000 |
| C | `피자` | 0 | 6 | 0.000 | 1 | 1.693 | 0.000 |

`피자`는 한 문서에만 있어 IDF는 크지만 B와 C에는 실제로 등장하지 않아 TF-IDF가 0입니다. 반대로 B의 `파스`는 IDF가 조금 작아도 짧은 B 문서에서 차지하는 비율인 TF가 커서 0.143이 됩니다. TF-IDF는 **드문 단어이면 무조건 높은 값**이 아니라 TF와 IDF가 모두 있어야 만들어지는 점수입니다.


## 예상하기

- `파스타`의 경계를 포함한 2-gram은 네 개이다.
- 세 문서 중 한 문서에만 있는 `피자`의 IDF가 두 문서에 있는 `파스`의 IDF보다 크다.
- 신경망 문서에는 `뉴런`, 컴퓨터 비전 문서에는 `이미지`, 책임 있는 AI 문서에는 `책임`이 높은 특징어로 나타날 것이다.
- ‘파스타, 피자’를 좋아하는 가상 취향은 파스타·피자 메뉴와 가장 높은 유사도를 보인다.


## 활동 1. 움직이는 창으로 n-gram 만들기


### 코드 살펴보기


1. `padded_word`는 단어 앞뒤에 공백을 붙여 경계를 표시합니다.<br>
2. `range(len(padded_word) - 1)`은 두 칸짜리 창이 시작할 위치를 만듭니다.<br>
3. `padded_word[index : index + 2]`는 현재 위치부터 두 글자를 자릅니다.<br>
4. 화면에서 경계가 보이도록 공백을 `·`로 바꿉니다. 실제 계산에서는 공백을 그대로 씁니다.<br>
5. `_char_ngrams(example)`은 같은 과정을 2·3·4글자 창으로 반복하고 조각별 횟수를 셉니다.


In [1]:
import sys
from pathlib import Path

current_folder = Path.cwd().resolve()
for candidate in (current_folder, *current_folder.parents):
    if (candidate / "jupyter_course" / "notebook_support.py").is_file():
        PROJECT_ROOT = candidate
        break
else:
    raise RuntimeError(
        "프로젝트 폴더를 찾지 못했습니다. 프로젝트 최상위 폴더에서 "
        r".\.venv\Scripts\python.exe -m notebook 명령으로 다시 시작하세요."
    )

if str(PROJECT_ROOT) not in sys.path:
    sys.path.insert(0, str(PROJECT_ROOT))

from jupyter_course.notebook_support import course_setup

setup = course_setup(PROJECT_ROOT)
PROJECT_ROOT = setup["root"]
raw_rows = setup["rows"]
meal_df = setup["frame"]
data_source = setup["source"]
print("프로젝트 폴더:", PROJECT_ROOT)
print("데이터 출처:", data_source)
print("급식 행 수:", len(raw_rows))

from neis_meal_ai.recommender import _char_ngrams, _tfidf_similarity
from IPython.display import display
import pandas as pd

ngram_word = "파스타"
padded_word = f" {ngram_word} "
two_grams = [
    padded_word[index : index + 2].replace(" ", "·")
    for index in range(len(padded_word) - 1)
]
three_grams = [
    padded_word[index : index + 3].replace(" ", "·")
    for index in range(len(padded_word) - 2)
]

print("경계를 표시한 단어:", padded_word.replace(" ", "·"))
print("2-gram:", two_grams)
print("3-gram:", three_grams)

example = "치즈 파스타"
grams = _char_ngrams(example)
print("\n추천기가 만든 2~4글자 조각 일부:")
print(list(grams.items())[:15])


프로젝트 폴더: <프로젝트 폴더>
데이터 출처: 남악고 NEIS 예비 데이터
급식 행 수: 5
경계를 표시한 단어: ·파스타·
2-gram: ['·파', '파스', '스타', '타·']
3-gram: ['·파스', '파스타', '스타·']

추천기가 만든 2~4글자 조각 일부:
[(' 치', 1), ('치즈', 1), ('즈 ', 1), (' 치즈', 1), ('치즈 ', 1), (' 치즈 ', 1), (' 파', 1), ('파스', 1), ('스타', 1), ('타 ', 1), (' 파스', 1), ('파스타', 1), ('스타 ', 1), (' 파스타', 1), ('파스타 ', 1)]


### 결과 해석하기

두 칸짜리 창을 한 칸씩 옮겨 네 개의 2-gram을 만들었습니다. 추천기는 각 단어에 같은 방법을 적용하고, 2·3·4글자 조각을 모두 모읍니다.


## 활동 2. 작은 세 문서의 TF-IDF 계산표 만들기


### 코드 살펴보기


1. `tiny_document_rows`는 문서 이름과 내용을 함께 저장해 어느 결과가 어느 문서에서 나왔는지 잃지 않게 합니다.<br>
2. `tiny_counts`는 각 문서의 2~4글자 조각과 등장 횟수를 셉니다.<br>
3. 바깥쪽 `for`는 문서 A·B·C를, 안쪽 `for`는 `파스`·`피자`를 차례로 살핍니다.<br>
4. `count / total_ngrams`가 TF이고, `sum(compared_term in other ...)`가 DF입니다.<br>
5. `tf * idf`로 최종 TF-IDF를 구한 뒤 계산식까지 `계산` 열에 남깁니다.<br>
6. `pd.DataFrame(...)`은 여섯 줄의 계산 결과를 읽기 쉬운 표로 보여 줍니다.


In [2]:
import math

tiny_document_rows = [
    {"문서": "A", "내용": "파스타 피자"},
    {"문서": "B", "내용": "파스타"},
    {"문서": "C", "내용": "밥 국"},
]
tiny_documents = [row["내용"] for row in tiny_document_rows]
tiny_counts = [_char_ngrams(text) for text in tiny_documents]
terms_to_compare = ["파스", "피자"]

print("[비교할 작은 문서]")
display(pd.DataFrame(tiny_document_rows))

toy_tfidf_table = []
for row, counts in zip(tiny_document_rows, tiny_counts):
    total_ngrams = sum(counts.values())
    for compared_term in terms_to_compare:
        count = counts[compared_term]
        tf = count / total_ngrams
        df = sum(compared_term in other for other in tiny_counts)
        idf = math.log((len(tiny_documents) + 1) / (df + 1)) + 1
        tfidf = tf * idf
        toy_tfidf_table.append(
            {
                "문서": row["문서"],
                "조각": compared_term,
                "등장 횟수": count,
                "전체 조각 수": total_ngrams,
                "TF": round(tf, 3),
                "DF": df,
                "IDF": round(idf, 3),
                "TF-IDF": round(tfidf, 3),
                "계산": (
                    f"{count} ÷ {total_ngrams} × {idf:.3f} "
                    f"= {tfidf:.3f}"
                ),
            }
        )

print("[TF-IDF 계산표]")
display(pd.DataFrame(toy_tfidf_table))


[비교할 작은 문서]


,문서,내용
0,A,파스타 피자
1,B,파스타
2,C,밥 국


[TF-IDF 계산표]


,문서,조각,등장 횟수,전체 조각 수,TF,DF,IDF,TF-IDF,계산
0,A,파스,1,15,0.067,2,1.288,0.086,1 ÷ 15 × 1.288 = 0.086
1,A,피자,1,15,0.067,1,1.693,0.113,1 ÷ 15 × 1.693 = 0.113
2,B,파스,1,9,0.111,2,1.288,0.143,1 ÷ 9 × 1.288 = 0.143
3,B,피자,0,9,0.000,1,1.693,0.000,0 ÷ 9 × 1.693 = 0.000
4,C,파스,0,6,0.000,2,1.288,0.000,0 ÷ 6 × 1.288 = 0.000
5,C,피자,0,6,0.000,1,1.693,0.000,0 ÷ 6 × 1.693 = 0.000


### 결과 해석하기

A·B·C의 두 조각을 모두 계산해 여섯 줄의 표로 만들었습니다. 등장하지 않은 조각은 TF가 0이므로 IDF가 커도 TF-IDF는 0입니다. B의 `파스`는 짧은 문서 안에서 차지하는 비율이 커서 A의 `파스`보다 높은 값을 받습니다.


## 활동 3. A 문서의 ‘피자’ 점수를 한 줄씩 따라가기


### 코드 살펴보기


1. `term_count_in_a`는 TF의 분자, `all_ngram_count_in_a`는 분모입니다.<br>
2. TF 계산을 `1 ÷ 15`처럼 실제 숫자로 출력해 공식과 결과를 연결합니다.<br>
3. `term_df`는 `피자`가 있는 문서만 세므로 1입니다.<br>
4. 세 문서라는 뜻의 3과 DF 1을 IDF 식에 넣어 1.6931을 얻습니다.<br>
5. `term_tf_in_a * rare_idf`는 0.0667과 1.6931을 곱해 0.1129를 만듭니다.<br>
6. 마지막 줄은 두 문서에 나타난 `파스`의 IDF가 더 작은지 비교합니다.


In [3]:
term = "피자"
term_count_in_a = tiny_counts[0][term]
all_ngram_count_in_a = sum(tiny_counts[0].values())
term_tf_in_a = term_count_in_a / all_ngram_count_in_a
term_df = sum(term in counts for counts in tiny_counts)
rare_idf = math.log(
    (len(tiny_documents) + 1) / (term_df + 1)
) + 1
rare_tfidf_in_a = term_tf_in_a * rare_idf

common_df = sum("파스" in counts for counts in tiny_counts)
common_idf = math.log(
    (len(tiny_documents) + 1) / (common_df + 1)
) + 1

print("1. A 문서의 '피자' 등장 횟수:", term_count_in_a)
print("2. A 문서의 전체 2~4글자 조각 수:", all_ngram_count_in_a)
print(
    "3. TF =",
    f"{term_count_in_a} ÷ {all_ngram_count_in_a}",
    "=",
    round(term_tf_in_a, 4),
)
print("4. DF = '피자'를 포함한 문서 수 =", term_df)
print(
    "5. IDF =",
    f"log((3+1)/({term_df}+1))+1",
    "=",
    round(rare_idf, 4),
)
print(
    "6. TF-IDF =",
    f"{term_tf_in_a:.4f} × {rare_idf:.4f}",
    "=",
    round(rare_tfidf_in_a, 4),
)
print("\n비교: '파스'의 DF =", common_df, "· IDF =", round(common_idf, 4))


1. A 문서의 '피자' 등장 횟수: 1
2. A 문서의 전체 2~4글자 조각 수: 15
3. TF = 1 ÷ 15 = 0.0667
4. DF = '피자'를 포함한 문서 수 = 1
5. IDF = log((3+1)/(1+1))+1 = 1.6931
6. TF-IDF = 0.0667 × 1.6931 = 0.1129

비교: '파스'의 DF = 2 · IDF = 1.2877


### 결과 해석하기

A 문서의 `피자`는 1회, 전체 조각은 15개이므로 TF는 0.0667입니다. 세 문서 가운데 A에만 있으므로 DF는 1, IDF는 1.6931입니다. 두 값을 곱한 TF-IDF는 0.1129입니다. `파스`는 A와 B 두 문서에 있어 IDF가 1.2877로 더 작습니다.


## 활동 4. 한국어 문서 세 편의 TF-IDF를 문서별로 계산하기


### 코드 살펴보기


1. `sources.json`에서 파일명·제목·원문 SHA-256을 읽습니다.<br>
2. 운영체제마다 다른 줄바꿈을 `LF`로 통일한 뒤 `sha256(document_bytes).hexdigest()`로 지문을 만들고 기록된 값과 비교합니다.<br>
3. `re.findall(r"[가-힣]{2,}", text)`는 한글 두 글자 이상인 덩어리만 찾습니다.<br>
4. `Counter`는 문서 안에서 각 단어가 몇 번 나왔는지 세어 TF의 재료를 만듭니다.<br>
5. `document_summaries`는 문서별 글자 수, 분석 단어 수, 서로 다른 단어 수를 먼저 보여 줍니다.<br>
6. `calculated_rows`에는 단어마다 등장 횟수부터 TF-IDF까지 계산한 값을 모두 저장합니다.<br>
7. `ranked_rows`는 TF-IDF가 큰 순서로 정렬하고 앞의 여덟 단어를 선택합니다.<br>
8. `계산` 열은 `6 ÷ 297 × 1.6931 = 0.0342`처럼 실제 숫자가 공식에 들어가는 과정을 남깁니다.<br>
9. `cross_document_comparison`은 같은 단어가 세 문서에서 몇 번 나타나고 얼마의 점수를 받는지 나란히 보여 줍니다.<br>
10. 이 결과는 이 세 문서 안에서의 상대적인 특징이며, 단어의 절대적인 중요도를 판정한 것이 아닙니다.


In [4]:
from collections import Counter
from hashlib import sha256
import json
import math
import re

corpus_folder = PROJECT_ROOT / "data" / "tfidf_korean_documents"
manifest = json.loads(
    (corpus_folder / "sources.json").read_text(encoding="utf-8")
)

documents = []
verification_results = []
for source in manifest["documents"]:
    document_path = corpus_folder / source["file"]
    document_text = document_path.read_text(encoding="utf-8")
    document_text = document_text.replace("\r\n", "\n").replace("\r", "\n")
    document_bytes = document_text.encode("utf-8")
    actual_sha256 = sha256(document_bytes).hexdigest()
    is_verified = actual_sha256 == source["sha256"]
    verification_results.append(is_verified)
    if not is_verified:
        raise ValueError(f"원문 확인 실패: {source['file']}")
    documents.append((source["title"], document_text))
    print(
        source["title"],
        "→ 파일:", source["file"],
        "· 글자:", len(document_text),
        "· 원문 확인:", is_verified,
    )

stop_words = {
    "그리고", "그러나", "하지만", "또한", "대한", "대해", "위해",
    "통해", "있는", "있습니다", "있으며", "합니다", "됩니다",
    "입니다", "것을", "것이", "같은", "이러한", "다음", "가장",
    "사용", "사용할", "사용하여", "우리는", "여기서", "이를",
}

def count_korean_words(text):
    words = re.findall(r"[가-힣]{2,}", text)
    return Counter(word for word in words if word not in stop_words)

document_counts = [count_korean_words(text) for _, text in documents]
document_count = len(document_counts)
all_words = set().union(*(counts.keys() for counts in document_counts))
word_df = {
    word: sum(word in counts for counts in document_counts)
    for word in all_words
}

document_summaries = []
for (title, text), counts in zip(documents, document_counts):
    document_summaries.append(
        {
            "문서": title,
            "글자 수": len(text),
            "분석 단어 수": sum(counts.values()),
            "서로 다른 단어 수": len(counts),
        }
    )

print("\n[문서별 기본 정보]")
display(pd.DataFrame(document_summaries))

document_top_terms = {}
document_term_details = {}
all_document_term_rows = []
for (title, _), counts in zip(documents, document_counts):
    total_words = sum(counts.values())
    calculated_rows = []
    for word, count in counts.items():
        tf = count / total_words
        idf = math.log((document_count + 1) / (word_df[word] + 1)) + 1
        calculated_rows.append(
            {
                "단어": word,
                "등장 횟수": count,
                "전체 단어 수": total_words,
                "TF": tf,
                "DF": word_df[word],
                "IDF": idf,
                "TF-IDF": tf * idf,
            }
        )

    ranked_rows = sorted(
        calculated_rows,
        key=lambda row: (-row["TF-IDF"], row["단어"]),
    )[:8]
    detail_rows = []
    for rank, row in enumerate(ranked_rows, 1):
        detail = {
            "순위": rank,
            "단어": row["단어"],
            "등장 횟수": row["등장 횟수"],
            "전체 단어 수": row["전체 단어 수"],
            "TF": round(row["TF"], 4),
            "DF": row["DF"],
            "IDF": round(row["IDF"], 4),
            "TF-IDF": round(row["TF-IDF"], 4),
            "계산": (
                f"{row['등장 횟수']} ÷ {row['전체 단어 수']} "
                f"× {row['IDF']:.4f} = {row['TF-IDF']:.4f}"
            ),
        }
        detail_rows.append(detail)
        all_document_term_rows.append({"문서": title, **detail})

    document_term_details[title] = detail_rows
    document_top_terms[title] = [row["단어"] for row in detail_rows]
    print(f"\n[{title}] 상위 TF-IDF 계산표")
    display(pd.DataFrame(detail_rows))

comparison_words = ["뉴런", "이미지", "책임", "모델"]
cross_document_comparison = []
for compared_word in comparison_words:
    compared_df = word_df.get(compared_word, 0)
    compared_idf = math.log(
        (document_count + 1) / (compared_df + 1)
    ) + 1
    for (title, _), counts in zip(documents, document_counts):
        total_words = sum(counts.values())
        count = counts[compared_word]
        tf = count / total_words
        cross_document_comparison.append(
            {
                "단어": compared_word,
                "문서": title,
                "등장 횟수": count,
                "TF": round(tf, 4),
                "DF": compared_df,
                "IDF": round(compared_idf, 4),
                "TF-IDF": round(tf * compared_idf, 4),
            }
        )

print("\n[같은 단어를 세 문서에서 비교]")
display(pd.DataFrame(cross_document_comparison))

document_sources_verified = all(verification_results)


신경망 소개 → 파일: 01_신경망_소개.md · 글자: 2827 · 원문 확인: True
컴퓨터 비전 소개 → 파일: 02_컴퓨터_비전_소개.md · 글자: 5570 · 원문 확인: True
윤리적이고 책임 있는 AI → 파일: 03_책임_있는_AI.md · 글자: 2892 · 원문 확인: True

[문서별 기본 정보]


,문서,글자 수,분석 단어 수,서로 다른 단어 수
0,신경망 소개,2827,297,232
1,컴퓨터 비전 소개,5570,570,382
2,윤리적이고 책임 있는 AI,2892,391,310



[신경망 소개] 상위 TF-IDF 계산표


,순위,단어,등장 횟수,전체 단어 수,TF,DF,IDF,TF-IDF,계산
0,1,뉴런,6,297,0.0202,1,1.6931,0.0342,6 ÷ 297 × 1.6931 = 0.0342
1,2,신경망,7,297,0.0236,2,1.2877,0.0303,7 ÷ 297 × 1.2877 = 0.0303
2,3,입력,5,297,0.0168,1,1.6931,0.0285,5 ÷ 297 × 1.6931 = 0.0285
3,4,출력,4,297,0.0135,1,1.6931,0.0228,4 ÷ 297 × 1.6931 = 0.0228
4,5,모델,5,297,0.0168,2,1.2877,0.0217,5 ÷ 297 × 1.2877 = 0.0217
5,6,머신러닝,4,297,0.0135,2,1.2877,0.0173,4 ÷ 297 × 1.2877 = 0.0173
6,7,모델을,4,297,0.0135,2,1.2877,0.0173,4 ÷ 297 × 1.2877 = 0.0173
7,8,뉴런의,3,297,0.0101,1,1.6931,0.0171,3 ÷ 297 × 1.6931 = 0.0171



[컴퓨터 비전 소개] 상위 TF-IDF 계산표


,순위,단어,등장 횟수,전체 단어 수,TF,DF,IDF,TF-IDF,계산
0,1,이미지,18,570,0.0316,2,1.2877,0.0407,18 ÷ 570 × 1.2877 = 0.0407
1,2,이미지를,12,570,0.0211,1,1.6931,0.0356,12 ÷ 570 × 1.6931 = 0.0356
2,3,컴퓨터,12,570,0.0211,2,1.2877,0.0271,12 ÷ 570 × 1.2877 = 0.0271
3,4,광학,9,570,0.0158,1,1.6931,0.0267,9 ÷ 570 × 1.6931 = 0.0267
4,5,비전,6,570,0.0105,1,1.6931,0.0178,6 ÷ 570 × 1.6931 = 0.0178
5,6,탐지,6,570,0.0105,1,1.6931,0.0178,6 ÷ 570 × 1.6931 = 0.0178
6,7,색상,5,570,0.0088,1,1.6931,0.0149,5 ÷ 570 × 1.6931 = 0.0149
7,8,움직임,5,570,0.0088,1,1.6931,0.0149,5 ÷ 570 × 1.6931 = 0.0149



[윤리적이고 책임 있는 AI] 상위 TF-IDF 계산표


,순위,단어,등장 횟수,전체 단어 수,TF,DF,IDF,TF-IDF,계산
0,1,책임,6,391,0.0153,1,1.6931,0.0260,6 ÷ 391 × 1.6931 = 0.0260
1,2,결정을,5,391,0.0128,1,1.6931,0.0217,5 ÷ 391 × 1.6931 = 0.0217
2,3,데이터를,6,391,0.0153,2,1.2877,0.0198,6 ÷ 391 × 1.2877 = 0.0198
3,4,모델이,4,391,0.0102,1,1.6931,0.0173,4 ÷ 391 × 1.6931 = 0.0173
4,5,어떤,4,391,0.0102,1,1.6931,0.0173,4 ÷ 391 × 1.6931 = 0.0173
5,6,있다는,4,391,0.0102,1,1.6931,0.0173,4 ÷ 391 × 1.6931 = 0.0173
6,7,이는,5,391,0.0128,2,1.2877,0.0165,5 ÷ 391 × 1.2877 = 0.0165
7,8,중요한,5,391,0.0128,2,1.2877,0.0165,5 ÷ 391 × 1.2877 = 0.0165



[같은 단어를 세 문서에서 비교]


,단어,문서,등장 횟수,TF,DF,IDF,TF-IDF
0,뉴런,신경망 소개,6,0.0202,1,1.6931,0.0342
1,뉴런,컴퓨터 비전 소개,0,0.0000,1,1.6931,0.0000
2,뉴런,윤리적이고 책임 있는 AI,0,0.0000,1,1.6931,0.0000
3,이미지,신경망 소개,2,0.0067,2,1.2877,0.0087
4,이미지,컴퓨터 비전 소개,18,0.0316,2,1.2877,0.0407
5,이미지,윤리적이고 책임 있는 AI,0,0.0000,2,1.2877,0.0000
6,책임,신경망 소개,0,0.0000,1,1.6931,0.0000
7,책임,컴퓨터 비전 소개,0,0.0000,1,1.6931,0.0000
8,책임,윤리적이고 책임 있는 AI,6,0.0153,1,1.6931,0.0260
9,모델,신경망 소개,5,0.0168,2,1.2877,0.0217


### 결과 해석하기


#### 왜 이 단어의 점수가 높을까?

세 문서의 실제 1위는 다음과 같습니다.

| 문서 | 1위 단어 | 등장 횟수 | TF | DF | IDF | TF-IDF |
|---|---|---:|---:|---:|---:|---:|
| 신경망 소개 | 뉴런 | 6 | 0.0202 | 1 | 1.6931 | 0.0342 |
| 컴퓨터 비전 소개 | 이미지 | 18 | 0.0316 | 2 | 1.2877 | 0.0407 |
| 윤리적이고 책임 있는 AI | 책임 | 6 | 0.0153 | 1 | 1.6931 | 0.0260 |

`이미지`는 두 문서에 있어 IDF가 `뉴런`이나 `책임`보다 작습니다. 그래도 컴퓨터 비전 문서에서 18번 나타나 TF가 크기 때문에 세 1위 가운데 가장 높은 0.0407이 됩니다. 이처럼 최종 순위는 IDF 하나가 아니라 TF와 IDF의 곱으로 정해집니다.


## 활동 5. 가상 취향과 다섯 메뉴를 점수 순서로 비교하기


### 코드 살펴보기


1. `query`는 실제 학생 정보가 아닌 가상 취향 문장입니다.<br>
2. `tolist()`는 표의 메뉴 열을 문장 목록으로 바꿉니다.<br>
3. `_tfidf_similarity`는 취향과 각 메뉴의 TF-IDF 벡터 방향 가까움을 한 번에 계산합니다.<br>
4. `direct_hits`는 `파스타`, `피자`, `면`이 메뉴명에 그대로 들어 있는지 확인합니다.<br>
5. `common_ngrams`는 완전한 단어가 같지 않아도 함께 나타난 2~4글자 조각을 셉니다.<br>
6. `sort(...)`는 유사도가 큰 메뉴부터 정렬하고, `순위` 열에는 1부터 번호를 붙입니다.<br>
7. 마지막 표에서 어느 날짜가 얼마의 점수를 받았고 어떤 메뉴였는지 한 줄씩 확인합니다.


In [5]:
query = "파스타 피자 면"
menu_texts = meal_df["menu_text"].tolist()
similarity_array = _tfidf_similarity(menu_texts, query)
similarities = [round(float(value), 3) for value in similarity_array]

query_terms = query.split()
query_ngram_set = set(_char_ngrams(query))
menu_similarity_ranking = []
for date, score, menu in zip(meal_df["date"], similarities, menu_texts):
    menu = str(menu)
    direct_hits = [term for term in query_terms if term in menu]
    common_ngrams = query_ngram_set.intersection(_char_ngrams(menu))
    menu_similarity_ranking.append(
        {
            "날짜": str(date),
            "유사도": score,
            "직접 포함 단어": ", ".join(direct_hits) if direct_hits else "없음",
            "공통 글자 조각 수": len(common_ngrams),
            "메뉴": menu,
        }
    )

menu_similarity_ranking.sort(
    key=lambda row: (-row["유사도"], row["날짜"])
)
for rank, row in enumerate(menu_similarity_ranking, 1):
    row["순위"] = rank

ranking_columns = [
    "순위", "날짜", "유사도", "직접 포함 단어", "공통 글자 조각 수", "메뉴"
]
print("[가상 취향과 메뉴의 TF-IDF 유사도 순위]")
display(pd.DataFrame(menu_similarity_ranking)[ranking_columns])

chapter_result = {
    "chapter": "03",
    "similarities": similarities,
    "query": query,
    "two_grams": two_grams,
    "common_idf": common_idf,
    "rare_idf": rare_idf,
    "document_count": document_count,
    "document_summaries": document_summaries,
    "document_term_details": document_term_details,
    "document_top_terms": document_top_terms,
    "document_sources_verified": document_sources_verified,
    "cross_document_comparison": cross_document_comparison,
    "menu_similarity_ranking": menu_similarity_ranking,
    "toy_tfidf_table": toy_tfidf_table,
}


[가상 취향과 메뉴의 TF-IDF 유사도 순위]


,순위,날짜,유사도,직접 포함 단어,공통 글자 조각 수,메뉴
0,1,2026-06-24,0.141,"파스타, 피자",9,양송이스프 미트볼로제파스타 노엣지콤비네이션피자 열대과일치즈샐러드 채소모둠피클 열무김...
1,2,2026-06-25,0.018,없음,1,친환경쌀밥 돈육애호박찌개 오이부추무침 묵은지닭볶음탕 심쿵햄전 열무김치 파인애플
2,3,2026-06-29,0.018,없음,1,친환경쌀밥 아욱국 족발&찐순대 배추겉절이 무말랭이무침 비빔막국수 피치에빠진코코
3,4,2026-06-26,0.000,없음,0,매콤떡갈비마요덮밥 감자호박된장국 구운버섯샐러드 바질크림츄볶이 배추김치 아이스구슬
4,5,2026-06-30,0.000,없음,0,친환경잡곡밥 돈육김치찌개 근대나물무침 감자채햄볶음 치킨까스 배추김치 샤인머스켓


### 결과 해석하기


2026-06-24 식단은 `미트볼로제파스타`와 `피자`를 직접 포함하고 공통 글자 조각도 많아 0.141로 1위입니다. 0.018처럼 작은 양수는 완전한 단어가 같지 않더라도 짧은 문자 조각 일부가 겹칠 때 나올 수 있습니다. 0.000은 이번 가상 취향과 비교할 공통 특징이 거의 없다는 뜻입니다.

유사도는 **텍스트 표현의 가까움**입니다. 0.141이 0.018보다 가상 취향과 더 비슷하다는 뜻이지, 14.1점짜리 음식이거나 더 건강하다는 뜻은 아닙니다.


## 탐구 활동

query의 단어를 ‘밥 국물’ 또는 자신이 고른 가상 취향으로 바꾸고 가장 높은 날짜를 찾으세요.

먼저 기본값으로 한 번 실행하세요. 그다음 표시된 값 하나만 바꾸고, 달라진 결과를 아래에 적습니다.


In [6]:
practice_query = "밥 국물"
practice_scores = _tfidf_similarity(meal_df["menu_text"].tolist(), practice_query)
best_index = int(practice_scores.argmax())
print("가상 취향:", practice_query)
print("가장 비슷한 메뉴:", meal_df.iloc[best_index]["menu_text"])
print("유사도:", round(float(practice_scores[best_index]), 3))


가상 취향: 밥 국물
가장 비슷한 메뉴: 친환경쌀밥 돈육애호박찌개 오이부추무침 묵은지닭볶음탕 심쿵햄전 열무김치 파인애플
유사도: 0.009


### 내가 본 변화

- 내가 바꾼 값:  
- 화면에서 달라진 것:  
- 내 설명:


## 확인 문제

1. `파스타`의 경계를 포함한 문자 2-gram 네 개를 적어 보세요.
2. TF의 분자와 분모는 각각 무엇인가요?
3. DF는 한 조각의 총 반복 횟수와 어떻게 다른가요?
4. DF가 작아지면 IDF가 커지는 까닭을 식의 나눗셈과 연결해 설명해 보세요.
5. TF-IDF는 어떤 두 값을 곱한 것인가요?
6. 실제 한국어 문서에서 불용어를 빼는 까닭은 무엇인가요?
7. 문서의 SHA-256을 확인하는 까닭은 무엇인가요?
8. TF-IDF 유사도가 높으면 반드시 맛있거나 건강하다는 뜻인가요?


## 정답과 해설


1. `·파`, `파스`, `스타`, `타·`입니다.<br>
2. 분자는 그 문서 안의 해당 조각 횟수, 분모는 그 문서의 전체 n-gram 수입니다.<br>
3. DF는 같은 문서 안에서 몇 번 반복됐는지가 아니라 그 조각을 포함한 문서가 몇 개인지를 셉니다.<br>
4. DF가 작아지면 `(전체 문서 수+1)/(DF+1)`의 분모가 작아져 나눈 값과 IDF가 커집니다.<br>
5. 한 문서 안에서의 비율인 TF와 여러 문서에서의 희소성인 IDF를 곱합니다.<br>
6. `있습니다`, `그리고`처럼 여러 글에 흔하지만 주제를 구별하는 데 도움이 적은 말을 제외하기 위해서입니다.<br>
7. 내려받은 파일이 준비할 때 확인한 원문과 같은지, 빠지거나 바뀌지 않았는지 확인하기 위해서입니다.<br>
8. 아닙니다. 입력한 취향 글자와 메뉴 글자의 특징이 비슷하다는 뜻뿐입니다.


## 핵심 정리

- 문자 n-gram은 n칸짜리 창을 한 칸씩 옮겨 만드는 연속 글자 조각이다.
- TF는 한 문서 안의 비율, DF는 그 조각을 포함한 문서 수다.
- IDF는 여러 문서에서 드문 조각에 더 큰 값을 준다.
- TF-IDF는 한 메뉴에서 자주 나오면서 전체에서는 드문 특징을 크게 본다.
- 출처가 다른 실제 문서도 같은 단어 분리 규칙과 TF-IDF 식으로 특징어를 비교할 수 있다.
- SHA-256은 내려받은 문서가 준비한 원문과 같은지 확인하는 파일 지문이다.
- 텍스트 유사도는 취향 표현의 가까움이지 정답이 아니다.

### 다음 장

04장에서는 벡터의 방향을 비교하고 숫자 특징이 비슷한 식단을 묶습니다.


In [7]:
import json
print("__CHAPTER_RESULT__=" + json.dumps(chapter_result, ensure_ascii=False))


__CHAPTER_RESULT__={"chapter": "03", "similarities": [0.141, 0.018, 0.0, 0.018, 0.0], "query": "파스타 피자 면", "two_grams": ["·파", "파스", "스타", "타·"], "common_idf": 1.2876820724517808, "rare_idf": 1.6931471805599454, "document_count": 3, "document_summaries": [{"문서": "신경망 소개", "글자 수": 2827, "분석 단어 수": 297, "서로 다른 단어 수": 232}, {"문서": "컴퓨터 비전 소개", "글자 수": 5570, "분석 단어 수": 570, "서로 다른 단어 수": 382}, {"문서": "윤리적이고 책임 있는 AI", "글자 수": 2892, "분석 단어 수": 391, "서로 다른 단어 수": 310}], "document_term_details": {"신경망 소개": [{"순위": 1, "단어": "뉴런", "등장 횟수": 6, "전체 단어 수": 297, "TF": 0.0202, "DF": 1, "IDF": 1.6931, "TF-IDF": 0.0342, "계산": "6 ÷ 297 × 1.6931 = 0.0342"}, {"순위": 2, "단어": "신경망", "등장 횟수": 7, "전체 단어 수": 297, "TF": 0.0236, "DF": 2, "IDF": 1.2877, "TF-IDF": 0.0303, "계산": "7 ÷ 297 × 1.2877 = 0.0303"}, {"순위": 3, "단어": "입력", "등장 횟수": 5, "전체 단어 수": 297, "TF": 0.0168, "DF": 1, "IDF": 1.6931, "TF-IDF": 0.0285, "계산": "5 ÷ 297 × 1.6931 = 0.0285"}, {"순위": 4, "단어": "출력", "등장 횟수": 4, "전체 단어 수": 297, "TF": 0.0135, "DF